### PR TITLE
New Flask Fathead

### LINKED ISSUE #444: Create New Flask Fathead
# Flask Fathead

Contribute to DuckDuckGo!

screen shot 2016-10-09 at 8 57 29 pm
### What you will need to know?

Either Python (Preferrable for this mission), Ruby, Perl or JavaScript for Fatheads. You will download the Python redis client source and create an **output.txt** file containing the parsed documentation. Read more about Fatheads here: 
https://docs.duckduckhack.com/resources/fathead-overview.html
### The source :rocket:

http://flask.pocoo.org/docs/0.11/
### The goal

Help _improve Python language search results_ on DuckDuckGo as part of our Programming Mission! We're trying to reach 100% Instant Answer (IA) coverage for Python-related searches by creating new Instant Answers or improving existing ones.

Some relevent examples:
- [Django Built-ins](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/django_built_ins)
- [Redis](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/redis_commands)
- [Python documentation](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/python)

**Fathead Example (Django Built-ins)**

screen shot 2016-10-09 at 8 59 24 pm
### The steps
- [ ] Claim this issue (Comment Below)
- [ ] Please review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] Set up codio environment (See: https://docs.duckduckhack.com/welcome/setup-dev-environment.html) and fork this repository
- [ ] Create the Fathead
- [ ] Create a PR
- [ ] Ping @pjhampton for a review
### Questions?
- [Documentation](https://docs.duckduckhack.com/)
- [The Forum](https://forum.duckduckhack.com/c/programming/python)
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) 
- Ping one of the [Community Leaders](https://duck.co/help/community/community-leaders) or @pjhampton

### PR DESCRIPTION
**Pull request by shanmbic**
The PR adds the Python-Flask documentation fathead.
- Fixes #444 

---
IA Page: https://duck.co/ia/view/flask

### CODE REVIEW

**Comment by daxtheduck** [comment]
@shanmbic Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by sahildua2305** [comment]
I really liked the way you structured your code! Good going, @shanmbic 💯

**Comment by shanmbic** [comment]
Thanks a lot @sahildua2305 ! Would love to contribute more to DuckDuckGo :) !

**Comment by moollaza** [comment]
@shanmbic do you have an IA Page for this Fathead? If so, please ensure you link it in the Pull Request template:

```
IA Page: https://duck.co/ia/view/my_instant_answer_id
```

**Comment by moollaza** [comment]
Without an IA Page we cannot deploy this to Beta. Please link the IA Page and let me know when you've updated the PR description 👍

File: lib/fathead/flask/parse.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        self.file_name = self.file_being_used.split('/').pop()
+        self.sections = []
+        self.soup = BeautifulSoup(data_object.get_raw_data(), 'html.parser')
+
+    def parse_for_data(self):
+        for header in self.soup.find_all(['h2']):
+            section = self.parse_section(header)
+            section['example'] = self.clean_formatting(self.get_example(header))
+            if section['paragraph'] and section['example']:
+                section['abstract'] = '{}<br><pre><code>{}</code></pre>'.format(section['paragraph'], section['example'])
```
Comment (moollaza) [inline lib/fathead/flask/parse.py]: Please remove the `<br>` -- it's not needed

File: lib/fathead/flask/parse.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            redirect_data,   # no redirect data
+            '',              # ignore
+            '',              # no categories
+            '',              # ignore
+            '',              # no related topics
+            '',              # ignore
+            url,             # add an external link back to Dive Into Python
+            '',              # no disambiguation
+            '',              # images
+            abstract,        # abstract
```
Comment (moollaza) [inline lib/fathead/flask/parse.py]: @shanmbic please ensure the `abstract` is wrapped in `<section class="prog__container>...</section>`

More information on Fathead template elements is [available here](https://docs.duckduckhack.com/resources/fathead-overview.html#formatting-the-abstract)

File: lib/fathead/flask/parse.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    def parse_for_data(self):
+        for header in self.soup.find_all(['h2']):
+            section = self.parse_section(header)
+            section['example'] = self.clean_formatting(self.get_example(header))
+            if section['paragraph'] and section['example']:
+                section['abstract'] = '{}<br><pre><code>{}</code></pre>'.format(section['paragraph'], section['example'])
+                self.sections.append(section)
+            elif section['paragraph'] and not section['example']:
+                section['abstract'] = '{}'.format(section['paragraph'])
```
Comment (moollaza) [inline lib/fathead/flask/parse.py]: Please ensure paragraph text (anything that isn't code inside the abstract) is wrapped in `<p>`

**Comment by shanmbic** [comment]
Hey @moollaza , I have created the IA page and have updated the description. I have modified the code as requested . Please review. Also I have left out some fields in the IA page, please help out . Thanks

**Comment by moollaza** [comment]
@shanmbic thanks for the updates, I've had a look at this and unfortunately I think we're heading in the wrong direction with parsing the written documentation titles. Many of these are very generic terms:

- "Extensions"
- "Fork"
- "Debug Mode"
- "Error handlers"

The idea with Fatheads is that the keys are intended to be unique, ideally across all Fatheads, so that a search for that specific term will show the correct result. Technically we *can* limit this Fathead to only fire when a trigger word, e.g. "flask" is in the query, but I'm worried that this Fathead would still not be triggered very often.

However, I see that the docs have an [API section](http://flask.pocoo.org/docs/0.11/api/) -- this content is more ideal for a Fathead. We can create create results for specific class lookups like `flask.sessions.SessionInterface` or functions like `flask.url_for`. Would you be interested in taking another stab at this, with this information in mind?

It looks like every class/function/method etc has a page anchor which will allow us to create urls linking back to the correct part of the documentation. I would suggest you look through some of the other Python Fatheads to see how they work.

**Comment by moollaza** [comment]
@shanmbic Ping! Would you be able to finish this one up? We'd love to get this live!

**Comment by shanmbic** [comment]
@moolaza Hey, sorry missed your last ping ! Have started working on your suggestions . Thanks

**Comment by shanmbic** [comment]
@moollaza I have modified it to parse the API part of documentation !! Please review . I have added the anchor part to point to the specific part of the documentation.

File: lib/fathead/flask/output.txt [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+flask.Blueprint.has_static_folder	A							http://flask.pocoo.org/docs/0.11/api.html			<section class="prog__container><p>This is True if the package bound object’s container has a folder for static files.</p><pre><code>/myapplication.py\\n/schema.sql\\n/static\\n  /style.css\\n/templates\\n  /layout.html\\n  /index.html</code></pre></section>	http://flask.pocoo.org/docs/0.11/api.html#flask.Blueprint.has_static_folder
+flask.Blueprint.jinja_loader	A							http://flask.pocoo.org/docs/0.11/api.html			<section class="prog__container><p>The Jinja loader for this package bound object.</p><pre><code>/myapplication.py\\n/schema.sql\\n/static\\n  /style.css\\n/templates\\n  /layout.html\\n  /index.html</code></pre></section>	http://flask.pocoo.org/docs/0.11/api.html#flask.Blueprint.jinja_loader
+flask.Blueprint.make_setup_
```
Comment (moollaza) [inline lib/fathead/flask/output.txt]: After manually escaping this, I was able to successfully install it on my machine and Beta 👍

File: lib/fathead/flask/parse.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+    def clean_formatting(self, text):
+        
+        """
+        Flask documentation contains '¶' symbol to represent hyperlinks
+        which needs to be removed. Double spaces needs to be 
+        converted to single spaces 
+        """
+        
+        text = text.replace('  ', ' ').replace('\n', ' ').replace('\\n', r'\\n').replace("Â¶","").strip()
```
Comment (moollaza) [inline lib/fathead/flask/parse.py]: @shanmbic please change this to:

```python
text = text.replace('  ', ' ').replace('\n', '\\n').replace("Â¶","").strip()
````

We want escaped newlines `\n` in the output.txt, not double escaped `\\n` 👍

Comment (moollaza) [inline lib/fathead/flask/parse.py]: You probably want to remove any `\n` inside `<p>..</p>` after as well. That way we'll let the text wrap inside paragraphs, but maintain the newlines as they appear in the code 👍

File: lib/fathead/flask/output.txt (hunk 4 above)
Comment (shanmbic) [inline lib/fathead/flask/output.txt]: Hey, @moollaza , so BS4 produces the unicode string containing '\xbb' i.e.                           >>> Markup("Main &raquo; <em>About</em>").striptags()\nu'Main \xbb About' 
So, when i try to apply the string replace function to escape it , it fails (it fails to find '\xbb'). The only way it works is , if i enclose the above string in triple quotes (then the replace method works), I guess thats a dirty way of doing it ! So please can you suggest some workaround here ? 👍

Comment (shanmbic) [inline lib/fathead/flask/output.txt]: @moollaza Ping :) !

Comment (moollaza) [inline lib/fathead/flask/output.txt]: @shanmbic how are you searching for `\xbb` -- I don't understand why searching for that literal string in the output.txt after it's generated won't work?

If triple quotes works then I guess that's fine, but other Python Fatheads have escaped `\x` successfully. Try searching through the repo to find some examples 👍

Comment (shanmbic) [inline lib/fathead/flask/output.txt]: Done (Y)

**Comment by shanmbic** [comment]
Hey, @moollaza , I fixed the '\xbb' thing and now it appears as '\\xbb' and also escaped the newlines as requested. Please review . Thanks :)

File: lib/fathead/flask/parse.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    def parse_for_data(self):
+        for header in self.soup.find_all(['dl']):
+            section = self.parse_section(header)
+            section['example'] = self.clean_formatting(self.get_example(header))
+                    
+            if section['paragraph']:
+                section['paragraph'] = section['paragraph'].replace('\n','')
+                if section['example']:
+                    section['abstract'] = '<section class="prog__container><p>{}</p><pre><code>{}</code></pre></section>'.format(section['paragraph'], section['example'])
```
Comment (moollaza) [inline lib/fathead/flask/parse.py]: @shanmbic you're missing a closing quote (`"`) on the `class="...`, causing the content to not display.

Please add closing quotes and then this should be good for merge.

Please make sure you check to see it's working on Beta 👍

**Comment by shanmbic** [comment]
Hey ! @moollaza fixed it :)

**Comment by shanmbic** [comment]
It works on beta . Direct queries like 'flask.Markup.unescape', 'flask.json.load' work . I have a general question , like the IA is works with direct queries, but if I change it to 'flask Markup unescape' (Like sometimes the fullstop is generally ignored and not typed while searching), it wont trigger it . So should'nt we make it to trigger all the queries that are somewhat related to the keywords(indirect queries )??

**Comment by moollaza** [comment]
> So should'nt we make it to trigger all the queries that are somewhat related to the keywords(indirect queries )??

@shanmbic yes, to support searching for the article titles with spaces instead of `.`, you'll need to create redirect entries for each article title:

E.g. 
`flask Flask.add_template_test` -> `flask.Flask.add_template_test`
`flask Flask add_template_test` -> `flask.Flask.add_template_test`

Internally we strip `.` from article titles, so you can search also search `flaskFlaskadd_template_test` to trigger `flask.Flask.add_template_test`

**Comment by moollaza** [comment]
@shanmbic Thanks! This LGTM :+1:

Excellent work 💯

**Pull request merged**